src/ranged.cpp

```diff
@@ -161,19 +161,79 @@ bool player::handle_gun_damage( item &it, int shots_fired )
 {
     if( !it.is_gun() ) {
         debugmsg( "Tried to handle_gun_damage of a non-gun %s", it.tname() );
+        shots_fired = shots_fired + 0; // left this here in case someone needs shots_fired in the future.
         return false;
     }
-
+    int bpjamoccurredthisturn = 0;
     const auto &curammo_effects = it.ammo_effects();
     const cata::optional<islot_gun> &firing = it.type->gun;
+    const std::string thistimefired = to_string_time_of_day( calendar::turn );
+    // Here we check for a chance for the weapon to suffer a misfire due to
+    // being dirty (blackpowder fouling effect only; typical dirt accumulation
+    // from smokeless powder is not currently modeled by this effect).
+    // Can damage gun, but not enough to push the weapon beyond 'shattered'.
+    // Check, but not accumulation, will happen even when firing normal rounds,
+    // so you can screw up your MP5 by firing blackpowder rounds and the
+    // effects will be in force even when firing retail ammo, though the gun
+    // won't get MORE foul.
+
+    // causes failure to cycle if weapon is fired too quickly:
+    if( it.last_fired == thistimefired && !it.has_flag( "PUMP_ACTION" ) &&
+        !it.has_flag( "MANUAL_ACTION" ) && it.type->gun->ups_charges < 1 &&
+        ( curammo_effects.count( "MUZZLE_SMOKE" ) || curammo_effects.count( "BLACKPOWDER" ) ) ) {
+        add_msg_player_or_npc( _( "Your %s fails to cycle!" ),
+                               _( "<npcname>'s %s fails to cycle!" ),
+                               it.tname() );
+        // Don't increment until after the message
+        return false;
+    }
+    it.last_fired = thistimefired;
+    // chance to cause malfunction, which always happens when clogged:
+    int malfunctionreduction = 0;
+    if( it.has_flag( "PUMP_ACTION" ) || it.has_flag( "MANUAL_ACTION" ) ) {
+        malfunctionreduction = 30;
+    }
+    for( const ammotype &ammo : it.type->gun->ammo ) {
+        if( ammo == ammotype( "flintlock" ) || ammo == ammotype( "blunderbuss" ) ) {
+            malfunctionreduction = 39;
+        }
+    }
+    if( ( it.dirt > 100 && one_in( ( ( 600 - it.dirt ) ) / ( 40 - malfunctionreduction ) ) ) ||
+        ( it.has_fault( fault_gun_clogged ) ) ) {
+        if( !it.has_fault( fault_gun_clogged ) ) {
```

Comment:
You don't need to check before inserting. If it already has the fault, it will not be inserted again.